design/login.md

```diff
@@ -0,0 +1,103 @@
+# login
+
+The login package handles DC/OS login related operations.
+
+## Goals
+
+The goals of the login package are to:
+
+- Detect available login providers for a DC/OS cluster.
+- Log in to a DC/OS cluster using a given login provider.
+
+## Login providers detection
+
+To detect available login providers for a given DC/OS cluster, the CLI issues a GET request to the
+`/acs/api/v1/auth/providers` endpoint. This endpoint returns a list of providers in JSON. Each
+top-level key refers to the login provider ID, while its value is the login provider definition:
+
+``` json
+{
+  "dcos-users": {
+    "authentication-type": "dcos-uid-password",
+    "client-method": "dcos-usercredential-post-receive-authtoken",
+    "config": {
+      "start_flow_url": "/acs/api/v1/auth/login"
+    },
+    "description": "Default DC/OS authenticator"
+  }
+}
+```
+
+The above definition is a default DC/OS login provider, its ID is **dcos-users**.
+The ID has no special semantic apart from uniquely identifying a login provider.
+
+When the `/acs/api/v1/auth/providers` endpoint is not present, the available login provider is
+detected by making an unauthenticated `HEAD` request to a well-known protected DC/OS resource and
+reading the `WWW-Authenticate` header DC/OS returns in the 401 response:
+
+- `acsjwt` : the cluster only supports the `dcos-users` provider as defined above.
+- `oauthjwt` : the cluster only supports the OpenID implicit flow provider, as defined below:
+
+``` json
+{
+  "dcos-oidc-auth0": {
+    "authentication-type": "oidc-implicit-flow",
+    "client-method": "browser-prompt-authtoken",
+    "config": {
+      "start_flow_url": "/login?redirect_uri=urn:ietf:wg:oauth:2.0:oob"
+    },
+    "description": "Google, GitHub, or Microsoft"
+  }
+}
+```
+
+## Log in to a DC/OS cluster
+
+Log in to a DC/OS cluster is the action of presenting some credentials to the cluster in order to
+get an authorization token in return (also known as `ACS token`).
+
+The DC/OS CLI supports 6 login providers types (`authentication-type`):
+
+- **dcos-uid-password** : Log in using a standard DC/OS user account (username and password).
+- **dcos-uid-password-ldap** : Log in using an LDAP user account (username and password).
+- **dcos-uid-servicekey** : Log in using a DC/OS service user account (username and private key).
+- **saml-sp-initiated** : Log in using SAML 2.0.
+- **oidc-authorization-code-flow** : Log in using OpenID Connect authorization code flow.
+- **oidc-implicit-flow** : Log in using OpenID Connect implicit flow.
+
+A login happens through the `dcos auth login` command, which accepts the following flags:
+
+- **--provider** (string) : Specify the login provider ID to use.
+- **--username** (string) : Specify the username.
+- **--password** (string) : Specify password in plaintext (insecure).
+- **--password-file** (string) : Specify the path to a file that contains the password (insecure).
```

Comment:
Why is reading a password from a file is considered insecure?

Reply:
The insecure part is not about the reading, but about having a password stored in plaintext in a file.

Reply:
Although files can be leaked through bad practices, I'd consider files with 0600 permissions to be adequately secure.  I'd be more worried that the support for storing passwords in environment variable will encourage users to add their password to their `.profile` file, where we cannot detect that the password is exposed.

Why is there no support for interactive "getpass"-like support for entering the password?

Reply:
There is, these options are for scripting purposes. If they are not passed the user is prompted for a password.